daemon/delete_test.go

```diff
@@ -4,40 +4,160 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"strings"
 	"testing"
 
 	"github.com/docker/docker/api/types"
 	containertypes "github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/container"
 )
 
-func TestContainerDoubleDelete(t *testing.T) {
+// TestContainerDeletePaused tests that a useful error message and instructions is given when attempting
+// to remove a paused container (#30842)
+func TestContainerDeletePaused(t *testing.T) {
 	tmp, err := ioutil.TempDir("", "docker-daemon-unix-test-")
 	if err != nil {
 		t.Fatal(err)
 	}
 	defer os.RemoveAll(tmp)
-	daemon := &Daemon{
+	d := &Daemon{
 		repository: tmp,
 		root:       tmp,
 	}
-	daemon.containers = container.NewMemoryStore()
+	c := &container.Container{
+		CommonContainer: container.CommonContainer{
+			ID:     "test",
+			State:  &container.State{Paused: true, Running: true},
+			Config: &containertypes.Config{},
+		},
+	}
+
+	d.containers = container.NewMemoryStore()
+	d.containers.Add(c.ID, c)
+
+	err = d.ContainerRm(c.ID, &types.ContainerRmConfig{ForceRemove: false})
+	if err == nil {
+		t.Fatal("expected error when removing container, got none")
+	}
 
-	container := &container.Container{
+	expected := "cannot remove a paused container"
+	if !strings.Contains(err.Error(), expected) {
+		t.Fatalf("expected err: %v, got %v", expected, err)
+	}
+
+	expected = "Unpause and then stop the container before attempting removal or force remove"
+	if !strings.Contains(err.Error(), expected) {
+		t.Fatalf("expected err: %v, got %v", expected, err)
+	}
+}
+
+// TestContainerDeleteRestarting tests that a useful error message and instructions is given when attempting
+// to remove a container that is restarting (#30842)
+func TestContainerDeleteRestarting(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "docker-daemon-unix-test-")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmp)
+	d := &Daemon{
+		repository: tmp,
+		root:       tmp,
+	}
```

Comment:
I think a function like this would get a lot of re-use in daemon unit tests:

```go

func newDaemonWithTmpRoot(t *testing.T) (*Daemon, func()) {
 	tmp, err := ioutil.TempDir("", "docker-daemon-unix-test-")
	assert.NilError(t, err)
 	d := &Daemon{
 		repository: tmp,
 		root:       tmp,
 	}
 	d.containers = container.NewMemoryStore()
 	return d, func() { os.RemoveAll(tmp) }
}

...

d, cleanup := newDaemonWithTmpRoot(t)
defer cleanup()
```

Even if these get changed to table tests this is a nice abstraction that will keep the test function clean.